### PR TITLE
Document edge cases in the test-suite

### PR DESCRIPTION
**Pull request by qaristote**
This PR implements tests to document edge cases of the backends, as described in #51.

### CODE REVIEW

File: smtlib-backends-process/tests/EdgeCases.hs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+-- respond, not even with an empty line.
+commandNoResponse :: IO ()
+commandNoResponse = checkNoResponse "(set-option :print-success false)"
+
+checkNoResponse :: Builder -> IO ()
+checkNoResponse cmd = Process.with Process.defaultConfig $ \handle -> do
+  let backend = Process.toBackend handle
+      stdout = getStdout $ Process.process handle
+  SMT.send_ backend cmd
+  responseAvailable <- hReady stdout
```
Comment (facundominguez) [inline smtlib-backends-process/tests/EdgeCases.hs:44]: This test is doomed to fail intermittently if the backend does respond. You would have to insert a delay to decrease the chance of the test passing when the response is coming late. There is no good delay to choose, as it depends on where the test is run, and tests would be artificially taking longer to run without doing any useful work.

Perhaps it would be fine to just document the behavior here without actually trying to execute it.

Comment (qaristote) [reply smtlib-backends-process/tests/EdgeCases.hs:44]: Another option is to send another command that produces an output such as `(check-sat)` and check that there's no extra new line at the beginning of the response.

Comment (qaristote) [reply smtlib-backends-process/tests/EdgeCases.hs]: https://github.com/tweag/smtlib-backends/pull/53/commits/0014c19ca589f9c31985daa0cb27292c7726ddbb

File: smtlib-backends-process/tests/EdgeCases.hs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+checkNoResponse :: Builder -> IO ()
+checkNoResponse cmd = do
+  Just response <- Process.with Process.defaultConfig $ \handle -> do
+    let backend = Process.toBackend handle
+    -- using 'SMT.send' instead would hang the program
+    SMT.send_ backend cmd
+    -- (check-sat) will produce "sat"
+    LBS.stripSuffix "sat" <$> SMT.send backend "(check-sat)"
+  assertBool ("expected no response, got: '" <> LBS.unpack response <> "'") $ LBS.null response
```
Comment (facundominguez) [inline smtlib-backends-process/tests/EdgeCases.hs:43]:
```suggestion
  assertEqual "expected no response" "" response
```
Uses `show`, which works without having to interpret the bytes as text. Also inserts surrounding double quotes.

File: smtlib-backends-z3/tests/EdgeCases.hs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+-- | Upon processing a command producing no output, the backend will respond
+-- with an empty output.
+commandNoResponse :: IO ()
+commandNoResponse = checkEmptyResponse "(set-option :print-success false)"
+
+checkEmptyResponse :: Builder -> IO ()
+checkEmptyResponse cmd = Z3.with Z3.defaultConfig $ \handle -> do
+  let backend = Z3.toBackend handle
+  response <- SMT.send backend cmd
+  assertBool ("expected empty response, got: '" <> LBS.unpack response <> "'") $ LBS.null response
```
Comment (facundominguez) [inline smtlib-backends-z3/tests/EdgeCases.hs:31]:
```suggestion
  assertEqual "expected no response" "" response
```

**Review by facundominguez — APPROVED** [review]
LGTM
